[SYCL][DOC] Spec for discard queue events (#4922)
intel/llvm: sycl/doc/extensions/Avoid_event_creation/avoid_event_creation.asciidoc

> = Avoid event creation

Reviewer: This should be the name of the extension, typically something like `SYCL_EXT_ONEAPI_NAME`

Reviewer: Bikeshedding: I think "avoid event creation" is a bad name, because "avoid" sounds like a weak suggestion (i.e. "the runtime should avoid creating events if it is able to do so").  The semantics you propose here are much stronger than that -- my interpretation is closer to "the runtime should never create events, even if a developer attempts to use a feature that would require them").

Suggestions:
- `SYCL_EXT_ONEAPI_DISCARD_EVENTS` (implies any events that would be created are discarded instead)
- `SYCL_EXT_ONEAPI_DISABLE_EVENT_CREATION` (implies events are not created at all)
- `SYCL_EXT_ONEAPI_DISABLE_EVENTS` (implies all functionality related to events is disabled)

My preference is "discard events" or "disable events" -- they're both 6 characters shorter than "avoid event creation", and more aligned with my understanding of the proposal.

Author: ok, we can change the name, but for now, I changed the name to SYCL_EXT_ONEAPI_DISCARD_EVENTS

:source-highlighter: coderay
:coderay-linenums-mode: table

// This section needs to be after the document title.
:doctype: book
:toc2:
:toc: left
:encoding: utf-8
:lang: en

:blank: pass:[ +]

// Set the default source code type in this document to C++,
// for syntax highlighting purposes.  This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

// This is necessary for asciidoc, but not for asciidoctor
:cpp: C++

== Notice

IMPORTANT: This specification is a draft.

Copyright (c) 2021 Intel Corporation. All rights reserved.

NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
used by permission by Khronos.

NOTE: This document is better viewed when rendered as html with asciidoctor.

> GitHub does not render image icons.

Reviewer: You can remove this; this was true before, but we've started formatting extensions for GitHub instead.

Reviewer: You should also add a section about a feature test macro.  See https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/Uniform/Uniform.asciidoc for an example.

Author: ok, done.

== Dependencies

This extension is written against the SYCL 2020 specification, Revision 3.

== Status

Working Draft

This is a preview extension specification, intended to provide early access to
a feature for review and community feedback. When the feature matures, this
specification may be released as a formal extension.

Because the interfaces defined by this specification are not final and are
subject to change they are not intended to be used by shipping software
products.

== Introduction

This extension adds `avoid_event_creation` property for `sycl::queue`, by using this property

> the application informs SYCL RT library that it doesn't care about the returned event,
> i.e the application doesn't use the returned event and just ignores it.

Reviewer: ```suggestion
the application informs a SYCL implementation that the returned event can be discarded.
```
?

Author: applied your suggestion.

> In that case, the RT library makes a decision about internal optimization - voiding

Reviewer: ```suggestion
In that case, the implementation makes a decision about internal optimization -- avoiding
```

Author: applied your suggestion.

the creation of low-level event from the backend. The property affects only kernel cases,
i.e. avoid creating low-level event for submission of kernels.
Performance improvement is the main motivation for using this property.

Below is a usage example:

[source,c++]
----
  // in order queue is required for avoid_event_creation
  sycl::queue Queue( {property::queue::in_order{}, property::queue::avoid_event_creation{}} );

  // some USM preparations ..

  // creating "null" events
  sycl::event e1, e2, e3;

  // returning "null" events from each submission function:
  e1 = Queue.parallel_for(NDRange, [=](nd_item<1> item){ do_smth1(); });

  e2 = Queue.single_task([=](){ do_smth2(); });

  e3 = Queue.submit([&](handler &CGH) { CGH.parallel_for(NDRange, [=](nd_item<1> item){ do_smth3(); }); });

  Queue.wait();
----

In the example above, the application doesn't use sycl events: `e1`, `e2`, `e3`
and is waiting for the end of work by `queue::wait()`. The returned events will be

> `null events` - means that low-level events from backend are not created for them.

Reviewer: ```suggestion
_null_ events - means that low-level events from backend are not created for them.

```

Author: applied your suggestion.

> `null event` is the event created by the default event constructor, i.e `null event`

Reviewer: ```suggestion
null event is the event created by the default event constructor.

```

Author: applied your suggestion.

> has the same behavior as if the event was created by the default constructor.

Reviewer: I think the behavior here is subtle and potentially dangerous.

If you keep this behavior, then I think you need to expand on the discussion here with examples and non-normative notes to help users reason about what will happen.

SYCL 2020 says that the default constructor:

> Constructs an event that is immediately ready. The event has no dependencies and no associated commands. Waiting on this event will return immediately and querying its status will return `info::event_command_status::complete`.

If I'm reading your proposal and SYCL 2020 correctly, that means the following code examples won't give compilation or run-time errors, but also won't do what the user expects:
```c++
sycl::event e1;
e1 = q.parallel_for(...);
e1.wait(); // Returns immediately (because e1 is in "complete" status) and does not wait for the kernel
```

```c++
sycl::event e1, e2;
e1 = q1.parallel_for(...);
e2 = q2.parallel_for(e1, ...); // Runs immediately (because e1 is in "complete" status)
e2.wait();
```

Have you considered whether it would be possible to create a special event, instead, or perhaps a new `info::event::command_execution_status`?  If the event returned by one of these special queues had its status set not to `complete` but to something like `invalid` instead, we could also throw an `errc::invalid_operation` if somebody tried to wait on the event or pass it as an argument to some sort of dependency-expressing function.

Author: @Pennycook, I originally thought it was a good idea to return the default event when we have the new property is set. But for cases like multiple queues seems it makes sense to have different behaviour for that. For wait() probably also. I updated the document.

== Limitations

If the queue has this property and for some reason an event needs to be created,
then an exception with the `errc::invalid_operation` error code will be thrown.

The reasons for throwing the exception:

- using streams, buffer/image accessors (excluding local accessors)
- using together with `enable_profiling` property
- using OOO queue

> - using https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/Assert/SYCL_ONEAPI_ASSERT.asciidoc[fallback assert feature]

Reviewer: I'd be tempted to rewrite each of these as standalone paragraphs, so that you can provide more detail about each of these cases.  For example:

"If a kernel that uses `stream` objects, `buffer` or `image` accessors is submitted to a `queue` created with the `avoid_event_creation` property, the implementation must throw an exception with the `errc::invalid_operation` error code."

"If a `queue` is created with the `avoid_event_creation` property, the property list must also include the `in_order` property and must not include the `enable_profiling` property.  If these conditions do not hold, the implementation must throw an exception with the `errc::invalid_operation` error code."

We might need to expand these cases.  And writing separate paragraphs for each would also make it easier to have different error codes for different cases.  I don't know if `errc::invalid_operation` is the best error code -- would be interested to hear what @gmlueck  thinks.

Reviewer 2: I was originally thinking that this new queue property would only indicate that the event returned from `queue::submit()`, `queue::parallel_for()`, etc. was the special "invalid" event described in the comment above.  I was not thinking that the operations you list here would be invalid.  For example, if the user creates the queue with this new property and then submits a command that uses a buffer accessor, that would still work.  Maybe the implementation would still need to create a low-level event internally, but this would be an implementation detail.

I recognize that this means an application may not see a performance benefit in these cases.  I would add a non-normative note to this effect, rather than proscribing an exception in these cases.

I'm interested in other people's thoughts on this, though.

If we choose the semantic I propose, I'd name the extension something like `EXT_ONEAPI_NO_QUEUE_EVENTS` and I'd rename the queue property to something like `ext::oneapi::property::no_events` (or `ext::oneapi::property::queue::no_events`) with a description like:

> An application can specify this property if it does not expect to use the `event` object that is returned by any of the `queue` member functions.  When this property is specified, member functions on `queue` still return an `event` object, but most uses of this event result in an exception.  The implementation may perform more efficiently when this property is specified.

I would then follow this up with a non-normative note like:

> The `ext::oneapi::property::no_events` property is most useful when commands submitted to the queue use USM pointers, rather than accessors, to access data.  In order to realize a performance advantage with this property in DPC++, all of the following must be true:
>
> * The `queue` must also be constructed with `property::queue::in_order`.
> * Commands submitted to the `queue` must not use buffer or image accessors.
> * Commands submitted to the `queue` must not use a `stream` object.
> * Commands submitted to the `queue` must not call the `assert` macro.

And then there would be a section describing in more detail which operations on `event` are not allowed.

BTW, we should choose the namespace of the property based on some precedent.  Are other extended properties in the `ext::oneapi::property` namespace directly, or do we add another namespace level like `ext::oneapi::property::queue`?

Reviewer 2: I happened to be looking at other specifications that define extended events, and it looks like we do normally add the extra namespace.  Therefore, I suggest the property name be `ext::oneapi::property::queue::no_events`.

Reviewer: > I'm interested in other people's thoughts on this, though.

I could be convinced either way, but I'm leaning towards the exception behavior.

If the main usage of this feature is for optimization, and using the feature "incorrectly" effectively disables the optimization, it would be very useful to communicate that to the user somehow.  In my opinion, "You turned this optimization on, but your code is incompatible with it" is a better user experience than the code silently not running any faster.  If a user encounters an error after turning the optimization on, they know they need to change their code to remove usage of all the features listed; if things behave as normal, they may conclude the optimization isn't useful and turn it off.

I'd be more inclined to accept the alternative if the user could opt in (via something like an environment variable) to DPC++ printing optimization warnings to the console under certain conditions.  That might deliver a similar user experience, and has the added advantage of being something we could generalize to other extensions.

Author: > If the main usage of this feature is for optimization, and using the feature "incorrectly" effectively disables the optimization, it would be very useful to communicate that to the user somehow.

So, for now, I'm also leaning towards the exception behavior, I temporarily updated the document to align it to this behavior.

Reviewer 2: There's something about the semantic in the current draft that makes me uneasy.  It seems like we are letting our particular implementation drive the API, which seems backwards.  What happens if we change our implementation in the future in a way that requires creating an internal event?  We would change the definition of this property to throw an exception in a new case?  Obviously, we can't do that because it would cause existing applications to fail.

To illustrate with a specific example, let's suppose in the future that we change our implementation of reductions to use an event internally.  (I'm not sure if this is realistic, but bear with me.)  We don't currently list reductions as one of the things that causes an exception for a "discard_events" queue, so we can't change our implementation to throw an exception in this case, otherwise existing code will break.  So, now we will have a case where the implementation still creates an internal low-level event, even for a "discard_events" queue.  If this happens, I suppose we would add a non-normative note to the spec saying that there is no performance benefit in this case.

Therefore, it seems like it would be better to just add the non-normative note now.  Since it's non-normative, we can add (or remove) conditions in the future as our implementation changes.  Adding an environment variable as @Pennycook suggests seems reasonable as well, so people can use this envvar as a debugging aid.

Reviewer 3: +1 For making this property a hint that user doesn't use events rather than a requirement for SYCL RT to not create low level event. It would also allows us to remove limitations(or rename them to "cases where this property has no an effect" implementation note.)

Reviewer 2: I was thinking about this more.  The thing I don't like is that the name of the feature and the description imply that queues created with this property won't create any *internal low-level* events.  Since this is an implementation detail, the user has no way to control this.  Instead, I think the feature needs to describe the *SYCL* features that the application promises not to use.

My proposal above is written this way.  The definition of the `ext::oneapi::property::queue::no_events` means that the application promises not to use the `event` that is returned by any of the `queue` member functions.

However, it seems like @Pennycook and @alexanderfle would like the property to be even more restrictive.  For example, they would like the property to mean that the application is disallowed from submitting commands that uses accessors.

I think we could word the extension that way, but then we should rename the feature because this property places many more limitations on the queue beyond just the use of the `event`.  In this case, I'd suggest the name `SYCL_EXT_ONEAPI_LIMITED_QUEUE` (or maybe `SYCL_EXT_ONEAPI_RESTRICTED_QUEUE`), and I'd name the event `ext::oneapi::property::queue::limited` (or `ext::oneapi::property::queue::restricted`).  I'd describe the property like this:

> When a `queue` is constructed with this property, some operations on the queue are limited and the application must avoid using these operations.  In exchange for these limitations, the implementation may be able to perform certain optimizations.  The exact limitations are listed below.

The spec would then enumerate the SYCL features that are disallowed for queues constructed this way:

* The application's use of any `event` object returned by the queue is limited.  Queue member functions that return an `event` still return an object, but the returned `event` is in an invalid state and most operations on this event result in an  exception.  These limitations are described more fully below.

* The application must not submit any command to the `queue` that uses a buffer or image accessor.  Applications should use USM pointers instead to reference data.  Attempting to submit such a command causes the implementation to throw a synchronous exception with the `errc::invalid` error code from the kernel invocation command (e.g. `parallel_for`).

* The application must not capture a `stream` object as a kernel argument for commands submitted to the queue.  Attempting to submit such a command causes the implementation to throw a synchronous exception with the `errc::invalid` error code from the kernel invocation command (e.g. `parallel_for`).

* The application must not call the `assert` macro from a command that is submitted to the queue unless the device has native support for assertions (as specified by `aspect::ext_oneapi_native_assert`).  Attempting to submit such a command causes the implementation to throw a synchronous exception with the `errc::invalid` error code from the kernel invocation command (e.g. `parallel_for`).

* A queue constructed this way is implicitly in-order, so the queue behaves as though the property `property::queue::in_order` was specified.

* Profiling cannot be enabled on a queue that is constructed this way.  Attempting to construct a queue with both `ext::oneapi::property::queue::limited` and `property::queue::enable_profiling` causes the implementation to throw an exception with the `errc::invalid` error code.

One thing to keep in mind is that we cannot add more restrictions later because that would cause existing applications to break.  If we do decide later that we need more restrictions, we'd need to create a new property like `ext::oneapi::property::queue::more_limited`.

Reviewer: @gmlueck, you've convinced me that going with the "no `sycl::event`s are returned to the user" semantics are safer for now.  I wouldn't be surprised if we _already_ depended on one of the forbidden features to express certain dependencies in our reduction implementation (accessors, maybe?), and there might be other corner-cases we don't know about.  As long as we add some way to get some optimization hints, I'm happy.

Aside/Bikeshedding: I still think "discard events" is a good name for your preferred semantics.  "no events" is too easily interpreted as "no low-level events", but "discard events" can be interpreted as "any events that are created are discarded before they reach the user".  "discard" chosen deliberately because the behavior here reminds me of [nodiscard](https://en.cppreference.com/w/cpp/language/attributes/nodiscard) and [discarded value expressions](https://en.cppreference.com/w/cpp/language/expressions#Discarded-value_expressions).

Reviewer 2: Just to be clear, @Pennycook, you prefer my original proposed semantic over the second one ("limited queue")?

Regarding the name, I doubt customers will think "no events" means anything about internal low-level events.  That's an internal implementation detail, and I suspect our customers are less in tune with our implementation than we are.  I don't feel that strongly, though.  I do think that the feature-test macro name should have the word "queue" in it, though, since this extension applies only to queue events.  Therefore `SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS` (or `SYCL_EXT_ONEAPI_NO_QUEUE_EVENTS `).

Reviewer: > Just to be clear, @Pennycook, you prefer my original proposed semantic over the second one ("limited queue")?

Yes, sorry for being unclear.

> I do think that the feature-test macro name should have the word "queue" in it, though, since this extension applies only to queue events. Therefore SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS (or SYCL_EXT_ONEAPI_NO_QUEUE_EVENTS ).

Good point.  `SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS` looks good to me.

Author: @gmlueck, @Pennycook,  Do I understand correctly that using a new property as a hint is more preferable/safer? 
i.e. if the user creates the queue with this new property and then submits a command that uses a buffer accessor, that will still work (It will not throw an exception) and there will no optimization, in that case, **the normal event** will be returned.  But in the case when the necessary conditions are present for optimization, DPC++ RT will do it and return _null_ event for this case. Is this proposal the best now?

Reviewer: > in that case, the normal event will be returned. But in the case when the necessary conditions are present for optimization, DPC++ RT will do it and return null event for this case.

This wasn't the behavior that I expected, at least.  I think we would still want to have the semantics that the user can never use the event that's returned from a queue for certain operations.  Otherwise, we bind the semantics of this proposal to the behavior of our current implementation (which @gmlueck has convinced me is a bad idea).  We don't want it to be possible for somebody to run some code correctly today (because our implementation happens to create events), but have it fail to run in the future (because we removed the events from the implementation internals).

In your example, where a user submits a command using a buffer accessor, I'd expect the implementation to create an event and hold onto it internally, but to still return a `sycl::event` with a status like `invalid`, `unknown`, `discarded` (or whatever we decide to call that),  All queue operations with one of these queues should return a `sycl::event` that triggers an exception if you call `wait()`, or pass the event to a function like `depends_on`.

Reviewer 2: I completely agree with @Pennycook's response.

I also think we should stop calling this the "null event", and we should definitely not say it is the same as the event returned from the default event constructor.  I think the name "invalid event" is better, but I'm open to other suggestions.

Author: Ok, I updated the doc, but I thought a lot about the implementation of an `invalid` event. 
if I understood your suggestions correctly, the `invalid` event is not a separate entity, that is just `sycl::event` that reflects an `invalid` state of the event which throws an exception if you interact with it.

And maybe we shouldn't discount returning the `default` event instead of the `invalid` event. Sure, the `default` event has its own disadvantages, but I'm worried about the fact that the `invalid` event is dependent on the implementation of the `sycl::event`.
A simple example, if we add a new method for `sycl::event`, or change the behavior of an existing method, then we will also have to check what we expect from the `invalid` event and if we need to update the behavior for it.

@Pennycook, @gmlueck, who will be responsible for verifying that the extension is still working correctly after the `sycl::event` is updated?
But if we returned the `default` event, we would not have this problem.

Reviewer: > And maybe we shouldn't discount returning the default event instead of the invalid event. Sure, the default event has its own disadvantages, but I'm worried about the fact that the invalid event is dependent on the implementation of the `sycl::event`.

I think the disadvantages of using the default event here outweigh the advantages of having to modify less of the specification by a significant margin.  We need the extension to be easy to use if we want people to use it, and we need it to be very rigorously defined if we want it to be discussed as a potential feature for the next SYCL specification.

> A simple example, if we add a new method for `sycl::event`, or change the behavior of an existing method, then we will also have to check what we expect from the invalid event and if we need to update the behavior for it.

This is already true of your current extension -- if somebody introduces a new functionality to `queue`, somebody will have to define whether it's compatible with the semantics of your queue property.  I don't think we can escape this.

> @Pennycook, @gmlueck, who will be responsible for verifying that the extension is still working correctly after the sycl::event is updated?

@intel/dpcpp-specification-reviewers monitor extensions that are being introduced, so an update to the behavior of `sycl::event` wouldn't take us completely by surprise.  Updates to this specification might be necessary, but it might also be possible to say that other specification depends on this one and define interactions with the `invalid` status there... I don't think there's a hard and fast rule here.

Now that I think about it, I wonder if it would be a good idea to split the `invalid` execution status out into a separate extension, so we can decouple things a little bit... It might be overcomplicating things a little bit, but given recent experience with trying to deprecate and tease apart different components of extensions the increased modularity might be worth it.  What do you think, @gmlueck?

Reviewer 2: > Now that I think about it, I wonder if it would be a good idea to split the invalid execution status out into a separate extension, so we can decouple things a little bit

This doesn't seem like the right thing to me.  It will be more difficult for users if we split this into two extensions because then there will be two feature-test macros that they need to use if they want to write portable code.  The concept of the "invalid" event seem tightly tied to the rest of this extension, so I think it should all be one extension.

Reviewer: > This doesn't seem like the right thing to me. It will be more difficult for users if we split this into two extensions because then there will be two feature-test macros that they need to use if they want to write portable code.

I don't think they'd need to check both macros.  If we said that this extension depended on an invalid event extension, the macro for this extension would only be defined if the invalid event extension was supported.  Users who only cared about invalid events would check one macro, and users who only cared about this extension could forgo checking the invalid events macro.

> The concept of the "invalid" event seem tightly tied to the rest of this extension, so I think it should all be one extension.

That's true right now, I'm just wondering if it would benefit from being generalized.  Operations that currently throw asynchronous errors could be defined to set the event in "invalid" status, for example...  Anyway, just a thought.

> *Using this extension is well suited for kernels that use USM only, since sycl accessors are used to build a dependency graph, in which case low-level events are needed.

Reviewer: "Well-suited" implies to me that the extension is still compatible with accessors, just poorly suited.  I think you want stronger wording here.  Maybe something like the below?

```suggestion
NOTE : This extension is only compatible with kernels using Unified Shared Memory (USM) pointers.  SYCL accessors are used to build a dependency graph that rely on events being created and not discarded.

```

Author: applied your suggestion.

== Version

Built On: {docdate} +
Revision: 1

== Issues

None.

== Revision History

[cols="5,15,15,70"]
[grid="rows"]
[options="header"]
|========================================
|Rev|Date|Author|Changes
|1|2021-11-09|Alexander Flegontov |*Initial public working draft*
|========================================

> //************************************************************************
> //Other formatting suggestions:
> //
> //* Use *bold* text for host APIs, or [source] syntax highlighting.
> //* Use +mono+ text for device APIs, or [source] syntax highlighting.
> //* Use +mono+ text for extension names, types, or enum values.
> //* Use _italics_ for parameters.
> //************************************************************************
> © 2021 GitHub, Inc.
> Terms
> Privacy
> Security
> Status
> Docs
> Contact GitHub
> Pricing
> API
> Training
> Blog
> About

Reviewer: ```suggestion
```

You can delete all of this.

Author: Done.
